include/execution.hpp

```diff
@@ -2171,7 +2173,233 @@ namespace std::execution {
   /////////////////////////////////////////////////////////////////////////////
   // [execution.senders.adaptors.split]
   namespace __split {
+
+    struct __operation_base {
+      __operation_base * __next_;
+
+      virtual void __notify() noexcept = 0;
+      virtual ~__operation_base() = default;
+    };
+
+    template <class _Variant>
+      class __receiver {
+        __operation_base &__op_state_;
+        stop_source &__stop_source_;
+        _Variant &__data_;
+
+      public:
+        template <__one_of<set_value_t, set_error_t, set_stopped_t> _Tag, class... _As>
+        friend void tag_invoke(_Tag __tag, __receiver&& __self, _As&&... __as) noexcept {
+          try {
+            using __tuple_t = __decayed_tuple<_Tag, _As...>;
+            __self.__data_.template emplace<__tuple_t>(__tag, (_As &&) __as...);
+          } catch (...) {
+            using __tuple_t = __decayed_tuple<set_error_t, exception_ptr>;
+            __self.__data_.template emplace<__tuple_t>(set_error, current_exception());
+          }
+
+          __self.__op_state_.__notify();
+        }
+
+        friend auto tag_invoke(get_env_t, const __receiver& __self) {
+          return make_env<get_stop_token_t>(__self.__stop_source_.get_token());
+        }
+
+        __receiver(__operation_base &__op_state,
+                   stop_source &__stop_source,
+                   _Variant &__data) noexcept
+          : __op_state_(__op_state)
+          , __stop_source_(__stop_source)
+          , __data_(__data) {
+        }
+      };
+
+    enum class __state_t { __created, __started, __completed };
+
+    template <class _Sender>
+      struct __sh_state : __operation_base {
+        template <class... _Ts>
+          using __bind_tuples =
+            __bind_front_q<
+              __variant,
+              tuple<set_stopped_t>,
+              tuple<set_error_t, exception_ptr>,
+              _Ts...>;
+
+        using __bound_values_t =
+          __value_types_of_t<
+            _Sender,
+            make_env_t<get_stop_token_t, stop_token>,
+            __bind_front_q<__decayed_tuple, set_value_t>,
+            __q<__bind_tuples>>;
+
+        using __variant_t =
+          __error_types_of_t<
+            _Sender,
+            make_env_t<get_stop_token_t, stop_token>,
+            __transform<
+              __bind_front_q<__decayed_tuple, set_error_t>,
+              __bound_values_t>>;
+
+        using __receiver = __receiver<__variant_t>;
+
+        stop_source __stop_source_{};
```

Comment:
You should be able to use `in_place_stop_source` here and avoid the heap allocation and ref-counting.

Reply:
It used to be `in_place_stop_source`, but as @ericniebler rightly [noted](https://github.com/brycelelbach/wg21_p2300_std_execution/pull/460#discussion_r806296821) that P2300 uses `stop_source`. To be on the same page, here's the train of thoughts that justified the usage of the `in_place_stop_source` initially. 
- If `r`'s receiver contract has already been satisfied, an `op_state` won't use stop source.
- Otherwise, `op_state` constructs a callback that uses the stop source. But the lifetime of this callback ends when `op_state` is notified. Since `op_state` shares ownership of `sh_state` (which owns the stop source), the lifetime of the stop source is not an issue. 

What was unclear to me is the case when the stop token is used outside of the `op_state` context, for instance:
```
stop_token token{};

{
  auto snd = ex::read(ex::get_stop_token) | ex::split();
  sync_wait(snd | then([&token](auto t){ token = t; }));
}

if (token.stop_requested()) { // illegal for in place version
  // ...
}
```

The in-place version is not allowed in the code above since any calls to the `stop_requested` must strongly happen before the beginning of the invocation of the destructor of the source. I haven't considered updating the wording. As I understand, to use `in_place_stop_source` in the reference implementation, we'd have to:
- Explicitly state that stop tokens retrieved from the split environment shouldn't outlive`op_state`s (or split sender).
- Change the type of the stop token/source/callback in the wording. 

I suggest we use `stop_token` for now since it matches the current wording. @ericniebler, @lewissbaker if you agree with the steps above, I'll create a separate issue.

Reply:
The requirements for get_stop_token() should already be requiring that the token is not used after the receiver it was obtained from has completion signalled.

Reply:
@lewissbaker's analysis here is correct. Sorry I gave you bad advice earlier.